### ISSUE TITLE
Unzip should limit the filename length to 260 characters on Windows, not 200

### ISSUE THREAD

**Issue by osechet** [conan-io/conan#311]
On Windows the MAX_PATH is defined as 260 characters (https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx) but conan's unzip function limits the filename to 200 characters.

**Comment by memsharded** [comment]
The difference had a reason. The thing is that unzip is typically done in the "source" folder. When building, the sources will be copied to the "build/sha_xxx" folder, where sha_xxx would be 40 chars. 
In any case, you are right, that limit can be incremented, up to 221 chars, I'll do it.

I've heard that win10 plans to remove that limit, hopefully soon, it would be the end of this nightmare...

**Comment by osechet** [comment]
It would be great if win10 removes the limit. However, conan will still have to keep the limit to preserve compatibility with previous versions of the os. Windows 7 will be supported by Microsoft until 2020 and I think a lot of companies will continue to work with it for a while.

**Comment by memsharded** [comment]
Yes, we know, we will have to live with it for a while but at least some light at the end of the tunnel.

About this issue, I think I remember the original reason for the number being 200 and not 221. It is leaving some space for build systems. It is very, very common that build systems add some extra subfolders inside ``build/sha_xxx``, e.g. cmake will add something like "nameoftarget.dir/Release" and then inside the ``*.obj`` files. So the probabilities for this failure are high, and most likely they produce much uglier messages and more difficult to debug problems, you know, it is better to fail early.

Have you hit the limit? Which build system are you using? I think it is better to leave the limit as-is, probably the error message could be improved or at least documented properly in code. I have just added to my codebase
```diff
+                # the limit is 200 to account for 40 chars of SHA in build/sha_xxx
+                # and 20 chars for build system extra subfolders (cmake mytarget.dir/debug, etc)
```
It will go in the next commit, soon in develop.

**Comment by osechet** [comment]
Yes I hit the limit. I'm trying to create a package for Qt. It's already working on Linux but on Windows the limit is blocking me. The maximum length in the Qt zip is 154. Event if I move the .conan directory to the root directory, I get a minimum root path length of 52 characters: c:\.conan\data\Qt\5.6.0\data\osechet\testing\source. With this I already break the limit of 200; and I still have to add the zip destination (qt-everywhere-opensource-src-5.6.0, yeah they like long path).
As soon as I have time, I'll try to use a temporary directory to unzip, rename... and make it works within the limit.

**Comment by memsharded** [comment]
Damn limit! If you have the Qt conan package in a repository, I would also like to try some solution. I thought that Boost would be the longer, but if Qt hits the limit, probably it deserves a built-in helper to do this task. Lets make it work in your package, then we will integrate it in conan, is that ok? Thanks!

**Comment by Blubbz0r** [comment]
Have hit the problem while trying to create a Qt package... Not sure what the most appropriate solution would be. Renaming would only be possible for the top-level folder that has been unzipped, since the build system is relying on the rest of the folder structure. The only thing that I could imagine right now is the possibility to create some kind of redirect in this case that allows you to unzip e.g. to C:\ directly.

**Comment by lasote** [comment]
We have implemented a short_paths mechanism in the released conan 0.10.0.
The docs are not released yet but you can read the 0.10 branch in github to try it, please give us feedback!

https://github.com/conan-io/docs/blob/release/0.10/reference/config_files.rst

**Comment by Blubbz0r** [comment]
Not sure why, but it is not working for me.

These are the steps I have taken:

* updated conan (```conan --version```shows Conan version 0.10.0)
* adjusted the short_paths.conf to include ```Qt/5.7.0@habrich/testing C:\conan\Qt´´´
* I have a simple conanfile.py for Qt for testing. This is basically what it looks like (I have the Qt .zip package already on my machine, so it doesn't need to be downloaded every time):
```
from conans import ConanFile
from conans.tools import unzip
import os
class QtConan(ConanFile):
    name = "Qt"
    version = "5.7.0"
    settings = "os", "compiler", "arch", "build_type"
    zip_name = "qt-everywhere-opensource-src-5.7.0"

    def source(self):
        unzip("C:\\Qt\\%s.zip" % self.zip_name)
        os.unlink("C:\\Qt\\%s.zip" % self.zip_name)
```
* running ```conan export habrich``` and ```conan test_package```

However, it is still using the %HOMEPATH%/.conan/data/Qt/5.7.0/habrich/testing/source folder for everything.

**Comment by memsharded** [comment]
I think that you don't need to unzip to C://. Just unzip to your current folder, it will be one with a short path (I am not saying that this would fix it, just some info)

**Comment by Blubbz0r** [comment]
The C:\\Qt\\%s.zip parameter to unzip is the existing location. That is where I have the downloaded Qt package. the second parameter of unzip should be responsible for deciding the destination and defaults to ".", which should be enough, right?

**Comment by lasote** [comment]
Yes, It works for me in Linux.
I think you can have a typo in your short_paths.conf, do you miss the ":" ?

```
Qt/5.7.0@habrich/testing: C:\conan\Qt
```

**Comment by Blubbz0r** [comment]
Yes, the missing colon was the culprit. The simple script above is working now. Will extent the script next.

**Comment by Blubbz0r** [comment]
A simple Qt build is working fine for me now. However, I would like to see another way of configuring the target path. As a creator of the Qt package, I know that every one of my Windows users will need the short_paths config, but still every single one of them needs to remember to configure the short path locally. Maybe some sort of configuration of a general short path on the side of the user and invocation of a build in that path from within the conanfile.py of Qt would be an appropriate solution?

**Comment by lasote** [comment]
> I know that every one of my Windows users will need the short_paths config, but still every single one of them needs to remember to configure the short path locally. 

The generated packages(binaries) have very long paths too?

**Comment by Blubbz0r** [comment]
At least not in the case of my Qt build. This would only apply to the scenario where there are no binaries in the remote and the user needs to build locally.

**Comment by memsharded** [comment]
Conan ``tools.unzip()`` function now takes the full 260 characters. If the path is too long when copying from source => build folder, it will warn. Then, the new solution for the windows path length limit released in 0.11 can be used, just add ``short_paths=True`` to the conanfile.py

Please upgrade to conan 0.11 and re-open if necessary. Thank you!

**Comment by Samantha258** [comment]
Long Path Tool help me a lot when i have an issue like file deleting or renaming the file. Also good to use if file name and file extension is too long.
